olm/operator-registry.md

```diff
@@ -0,0 +1,322 @@
+
+The final implementation goal of this enhancement involves this set of steps:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without $operatorName included
+    ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without latest version of $operatorName included
+    ex: `operator-registry delete-latest foo example.db`
+
+As a point of context, these operations all output new database files. The intent of creating these commands is to allow the creation of new container images based on historical context of previous versions. These commands will be wrapped in tooling (outside the scope of this enhancement) that will be run as part of build environments or for local development. We are not creating a way for these commands to be run from inside the context of a cluster.
```

Comment:
The database file is an implementation detail. Previously, this database was always generated by the data in a given namespace on quay.io and generated at runtime. Here, the database will be generated as part of building a container image (through a set of commands) and the image itself is stored in a given registry (quay.io, dockerhub, etc.). The database file itself is part of the filesystem of the image that actually runs the service that hosts the data in the database.

Outside the scope of this enhancement itself there is an initiative to build a set of commands that can generate these index images so that all a developer needs to maintain is the yaml itself to build a bundle (also outside the scope of this enhancement, see https://github.com/openshift/enhancements/pull/27 for more details on that). The index image described in this enhancement would be managed by a CI system that aggregates a set of bundle container images. To reproduce this image, CI would be able to point to a given set of bundle images and rebuild the index (and therefore the database) from scratch.

Reply:
> The database file is an implementation detail.

Just to be clear, this is part of the implementation details.

Reply:
Yes, but from the perspective of a source control owner the database file itself is an implementation detail that will not be exposed to the user. The database file itself is described in detail elsewhere in this doc.

Reply:
I should make sure I read further:
https://github.com/openshift/enhancements/pull/24/files#diff-cdcfafa680aa53239dc81d77c79e4404R259

Sorry

Reply:
> To reproduce this image, CI would be able to point to a given set of bundle images and rebuild the index (and therefore the database) from scratch.

Does that mean it is up to the CI owner to figure out how to maintain a list of bundle images, add to it over time, track changes, etc?

Usually when we build and release a container image, we require that the entire source is available in one place so that the build is reproducible. The current registry images fit that mold, in that all the manifests are kept in one git repo, and you just run a build command on that. It's easy to understand and implement.

If you're proposing a different kind of process, I think it should be fully described so we can make sure it will meet reproducibility requirements, and that it will provide a user experience that will be conducive to community adoption.

Reply:
> Usually when we build and release a container image, we require that the entire source is available in one place so that the build is reproducible. The current registry images fit that mold, in that all the manifests are kept in one git repo, and you just run a build command on that. It's easy to understand and implement.

This is actually not accurate as of today. The current registry images do not fit that mold. They aggregate all the data that currently exists in an app-registry repository on Quay.io. We are replacing that concept with container images that aggregate the data, but the principle is the same.

Reply:
Sorry, I'll be more specific. Today, when you *build a registry image*, all the source material is in one place. Here is a very simple example: https://github.com/mhrivnak/demo-operator-registry/tree/1.0

There is one source of truth about what goes into that container image.

Reply:
None of the operators that are shipped as part of openshift today use that older API or are built that way. The fact that OLM and operator-registry happen to have an API that allows that is largely a user testing workflow and not something that exists in any production pipeline. The current operators that are shipped with openshift are operator-registry pods that pull data from app-registry.

From my perspective, the database file is something that is an under the hood implementation detail. In the future, it is definitely within scope to even replace that database file with an entirely different format. By making that database file public, we are essentially creating another possible migration point.

As an aside, CI developers won't and shouldn't interact with these commands directly. Instead, they will interact with a higher level tool. I can detail some of the bits of that tool and how it might work in conjunction with operator-registry commands if that would help.

Reply:
I see. It's hard to know how people releasing their own operators outside Red Hat are doing so; are we sure they aren't building these images? It probably doesn't matter, but I wouldn't write it off entirely. It's an OLM feature, and it's a simple way to independently distribute operators for consumption by OLM.

I am not advocating for surfacing that database file. What I am advocating for is having a single source of truth, in version control, for what gets built into a container image. There are lots of ways to do it, and I agree that using the database file for that purpose would not be great.

So how will someone express, and track over time, what gets built into these images?

Reply:
Is your concern how we could reproduce the image from scratch? We should be able to point the operator-registry at an entire registry namespace and build it from that list. The whole purpose of this enhancement is to define a storage mechanism for the operators -- as in, the storage of these things in the registry IS the storage. That means these should be 100% reproducible by the process I described above.

It is a valid point that there is another dimension (the current time) that means rebuilding an older index image from scratch isn't solved by this problem. If what you are asking for is a release valve to ensure that we could rebuild an older index, we could definitely do that by just storing the historical manifest that can be generated by a hypothetical `operator-registry manifest` command that returns the list of image shas used to build it. Then each time the index is built you could just store the json blob that outputs in source control. But I do not believe that is (or should be) the default workflow for a CI pipeline -- it would be an extremely costly operation to rebuild these from scratch when they are sufficiently large. This is an optimization to solve that problem.

Reply:
How expensive? Are we pre-optimizing and make it harder for our partners?

Reply:
@shawn-hurley @ralphbean This optimization of not having to rebuild from scratch for each version for each update is the number 1 reason why we designed this this way. If we feel like this is not a problem we actually need to solve then having the cumulative index aggregation isn't actually very valuable.

Reply:
Part of this is just long-standing policy on how we build and release artifacts. They must be reproducible. That's a big part of the reason brew doesn't have internet access. If a customer sees a container image on their system and asks what's in it, we need to have the source and a way to reproduce the build. Having the ability to demonstrate the provenance of released artifacts is very important as a vendor.

More practically, products (and projects) are going to use this as the release artifact that ties together multiple operators at specific versions into one product release. When product X releases version 1.2.6, they will do so with a catalog image as the vehicle. They will need to have careful control of what gets included in that release. There needs to be an auditable trail of exactly what software is included, at which versions, from which container images, with which CVEs, and with which fixes to CVEs. Customers will know that if they have the 1.2.6 catalog and upgrade all their operators, certain bugs are fixed.

Specifically for Red Hat, errata get written against product versions. The 1.2.7 release of product X will have an advisory that says specifically what is fixed in that release. We need a release artifact that defines the 1.2.7 release and exactly what's in it.

As simple approach that would make this widely accessible is to maintain a copy of the data in a git repo in some form. The way a current operator registry image gets built meets that need, and perhaps that approach can be adapted to fit the goals of this enhancement. If you have an `operator-registry manifest` command, as you described, it could primarily manage a local representation of the data in some form that can be version controlled and lifecycled. That doesn't necessarily require rebuilding the whole index every time. Tools like `createrepo_c` (which is trying to accomplish something quite similar) have become efficient at modifying an existing repository based only on content that has changed.

Reply:
But that isn't how package repositories work today though. A yum repository is essentially a set of XML files that have pointers to the path for rpms along with some metadata about what the package is. The individual rpm includes the source of the installation.

We are essentially making the exact same distinction here. To use the same metaphor

index image == yum repo (which is the subject of this enhancement)
bundle image == rpm (which is the subject of https://github.com/openshift/enhancements/pull/27)

This index image we are attempting to define is nothing more than a set of pointers to some stuff that can be installed along with some helpful metadata to define how to get it at install time. The concept is actually heavily influenced by `createrepo_c` which defines a lot of the same concepts. I do not believe that it should be necessary to be able to reliably put the entire set of manifests in a single image for the sake of reproducibility -- I would question the value of how important it is to reproduce something of this nature anyway just because it is an image. I certainly see the value of being able to rebuild the content of this image itself by storing a file which is essentially a list of images it was built from, however.

In reality, if you want to reproduce the yum repository by storing the xml somewhere, you can absolutely do that now. But if someone was to delete an rpm that the yum repo referred to, that would be mostly meaningless since the repository itself is no longer usable.

Reply:
The yum analogy is a good one. In that case, the RPMs must be in a local directory at the time the metadata files are generated. Once the XML has been created, it gets moved around with all of the RPMs;  they stay together as one large release output.

If the RPMs were being retrieved from a live API somewhere, or perhaps even being retrieved from multiple API sources, that would fundamentally change the picture. Then there would need to be a way to snapshot, at a point in time, exactly what set of RPMs that live API was serving. To some extent we have that with brew and koji. They enable you to tag RPMs into a repository build or product release, and then take care to keep the resulting sets immutable. We could make something like that for bundle images. But just as brew is major overkill for most use cases, we likewise do not want to require all distributors of operators to have a sophisticated set of tooling and services to manage snapshots of content served by live APIs.

While the yum world has advanced tooling like brew, koji, and [Pulp](https://pulpproject.org) to help people manage curated repositories over time, there is also the simple and bullet-proof option of putting RPMs in a directory, running createrepo_c, and then keeping the resulting output together as one immutable and reproducible release. We need something that simple for operators.

Reply:
I still question why `git repo == safe place to reproducibly store things` and `container registry == dangerous non reproducible place to store things`. In both cases the source is stored and hosted in an API. If we need to ensure those things are secure and will always be available, then we have backup procedures in place to ensure that the content in the registry is maintained. If the concern is that these indexes can be built from multiple sources: that is true, it is possible. It makes developer workflows like "What happens when I add my package to this index and try to install?" possible. But in reality when a production environment is managing these things, the bundles should be stored in a single secure safe place -- a managed and hosted container registry.

We are defining exactly the thing you are describing: operator-registry will get the manifests onto a local repo, run something to insert that data into the repo, and then results in an immutable blob that can be published. That is literally what `createrepo_c` does. Can you point to me what differences there are?

Reply:
git is the canonical way to track the provenance of source material. Serving content over live APIs is not *dangerous*, but a container registry today won't tell you anything about what it was serving yesterday. At Red Hat we've built a lot of tooling to make repositories of container images trackable, immutable, promotable from qa -> stage -> prod, etc. That's all to establish trust via an audit trail. A live API itself is not a signable, auditable artifact, so we've had to create processes and tools around that.

Maybe it would help if you could elaborate on the curation mechanism you have in mind. Suppose a product owner wants to include a specific list of operators together in a catalog and keep them updated over time; where is that list kept, what does it look like, and how is it used? Put another way, what is the single source of truth for what should be put into the catalog, and how does that get used?

Reply:
This is a very good discussion with lots of good points made. There are a couple of things I'd like to follow up on:

> The fact that OLM and operator-registry happen to have an API that allows that is largely a user testing workflow and not something that exists in any production pipeline. 

This is how we build the upstream catalog, and the non-appregistry tools were built for more than just testing. The appregistry server was added later to support the existing pipeline, which used appregistry, because we determined it would not be possible to change course at the particular stage of 4.1 development.

The way @mhrivnak is describing the use of operator-registry is exactly what was originally envisioned for operator-registry.

But there are some important points missing from this that changed our approach (and my mind):

1. We got lots of feedback from users that managing a set of all versions of your manifests was hard. It was easy to make mistakes, get them out of sync, forget to update them, etc.
2. The embedded icons and the descriptions take up a good chunk of disk space, and get stored for each version of each operator.
3. There is community interest in storing yaml manifests as images (see: ORAS, CNAB, helm, etc).
4. The "one big directory" model is a little harder to support external contributors (we can't, for example, easily give a user access to sub-directory of a git repo).

Because of (3), we needed to consider images as a potential source of data.  Because of (1), it seemed like that might be a good way to resolve current user frustration, and we'd get (4) and (2) solved for free (even though there are more options for 2).

> Then there would need to be a way to snapshot, at a point in time, exactly what set of RPMs that live API was serving.

This snapshotting is essentially what we're doing with this proposal. If it wasn't called out, all of the references to images are digest-only. Availability is a concern, though, which I think is a good point.

> A live API itself is not a signable, auditable artifact, so we've had to create processes and tools around that.

OCI/Docker registries and their artifacts they host are auditable and signable (there are perhaps _too many_ standard ways to do this), so I think Kevin's question is reasonable. 

But you are making a good point that all of Red Hat's provenance tracking is git-based and likely will be for some time. 

I also think that, since the file exists, users will treat it as an "API". One of the ways they can do that is to store it in a git repo and track changes to it there, and then load it up via the registry server.

One nice thing is that we can actually do both 😄 

 - Since we need to migrate, we will continue to support "direct" bundles (not stored in images). We could choose to never drop support.
 - If it's valuable, we could provide "all-in-one" registry images, or tooling to make them. It's a straightforward process to take bundle images, extract them, and store them.

Reply:
This is a very good discussion with lots of good points made. There are a couple of things I'd like to follow up on:

> The fact that OLM and operator-registry happen to have an API that allows that is largely a user testing workflow and not something that exists in any production pipeline. 

This is how we build the upstream catalog, and the non-appregistry tools were built for more than just testing. The appregistry server was added later to support the existing pipeline, which used appregistry, because we determined it would not be possible to change course at the particular stage of 4.1 development.

The way @mhrivnak is describing the use of operator-registry is exactly what was originally envisioned for operator-registry.

But there are some important points missing from this that changed our approach (and my mind):

1. We got lots of feedback from users that managing a set of all versions of your manifests was hard. It was easy to make mistakes, get them out of sync, forget to update them, etc.
2. The embedded icons and the descriptions take up a good chunk of disk space, and get stored for each version of each operator.
3. There is community interest in storing yaml manifests as images (see: ORAS, CNAB, helm, etc).
4. The "one big directory" model is a little harder to support external contributors (we can't, for example, easily give a user access to sub-directory of a git repo).

Because of (3), we needed to consider images as a potential source of data.  Because of (1), it seemed like that might be a good way to resolve current user frustration, and we'd get (4) and (2) solved for free (even though there are more options for 2).

> Then there would need to be a way to snapshot, at a point in time, exactly what set of RPMs that live API was serving.

This snapshotting is essentially what we're doing with this proposal. If it wasn't called out, all of the references to images are digest-only. Availability is a concern, though, which I think is a good point.

> A live API itself is not a signable, auditable artifact, so we've had to create processes and tools around that.

OCI/Docker registries and their artifacts they host are auditable and signable (there are perhaps _too many_ standard ways to do this), so I think Kevin's question is reasonable. 

But you are making a good point that all of Red Hat's provenance tracking is git-based and likely will be for some time. 

I also think that, since the file exists, users will treat it as an "API". One of the ways they can do that is to store it in a git repo and track changes to it there, and then load it up via the registry server.

One nice thing is that we can actually do both 😄 

 - Since we need to migrate, we will continue to support "direct" bundles (not stored in images). We could choose to never drop support.
 - If it's valuable, we could provide "all-in-one" registry images, or tooling to make them. It's a straightforward process to take bundle images, extract them, and store them.